src/dune_rules/lib.ml

```diff
@@ -1115,6 +1153,30 @@ end = struct
         in
         Memo.Build.map res ~f:Option.some
     in
+    let requires =
+      let open Resolve.O in
+      match implements with
+      | None -> resolved >>= fun r -> r.requires
+      | Some vlib ->
+        let* vlib = vlib in
+        let* vlib_closure_parents = vlib_closure_parents in
+        if Id.Map.mem vlib_closure_parents vlib.unique_id then
+          let loc = Lib_info.loc info in
+          Error.vlib_in_closure ~loc ~impl:info ~vlib:vlib.info |> Resolve.fail
+        else
+          let* resolved = resolved in
+          let+ requires = resolved.requires in
+          List.filter requires ~f:(fun lib -> not (equal lib vlib))
```

Comment:
This filtering wasn't there before. Why do we need it?

Reply:
We allowed users to specify V in the requires, so if we just forbid it now we'll break their code. But we do need to treat it specially now wrt to the check.